<!DOCTYPE html>
<html>
 <head>
 <title>User's Guide - Tag-Seq utilities</title>
 <link rel="stylesheet" type="text/css" href="style/docs_style.css">
 <style type="text/css">
 .row { vertical-align: top; height:auto !important; }
 .list {display:none; }
 .show {display: none; }
 .hide:target + .show {display: inline; }
 .hide:target {display: none; }
 .hide:target ~ .list {display:inline; }
 @media print { .hide, .show { display: none; } }
 </style>
<h2>
 <a href="http://people.oregonstate.edu/~meyere/tools.html" title="Back to Tools page"
  style="text-decoration: none; color: black; margin-left: 2%;">Tools | </a>
 <a href="#top" title="Collapse all menus" style="text-decoration: none; color: black;">
 Analysis of gene expression using Tag-Seq (v2.0)</a>
 </h2>
 <hr align="left" style="margin-left: 2%;">
 </head>

 <body>
  <div class="row" style="font-size: 90%;">
  <a href="#about" class="hide" id="about" style="text-decoration: none; color: black; margin-left: 2%">
   <span title="Click to expand">
   About this guide</span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%">
   <span title="Click to collapse">
   About this guide</span></a>
   <div class="list">
   <p align="left" style="margin-left: 2%">This page outlines our standard pipeline for analysis of gene expression using Tag-Seq,
   a cost-effective and accurate approach. This guide covers the most recently updated version of the Tag-Seq utilities pipeline as
   of July 2018 (v2.0).<br><br>
   These tools are intended to be used on a high-performance computing cluster, and assume a basic knowledge of Linux and command-line
   tools. To analyze large numbers of samples in parallel, the user will probably want to (a) combine some of the
   following steps into shell scripts, and (b) submit each job to a job scheduler such as SGE. Since the details of
   those steps may vary from one cluster to another, we have left those details up to the end user.<br><br>
   Click on each header to expand or collapse that section.
   </p>
   </div>
  <br>
  </div>
  <div class="row">
  <a href="#install" class="hide" id="install" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <b><span title="Click to expand">
   &#9655 Download scripts and install required software</span></b></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <b><span title="Click to collapse">
   &#9665 Download scripts and install required software</span></b></a>
   <div class="list">
   <ul>
   <li><b>BioPerl</b></li>
    <p>Our scripts rely on BioPerl modules. First, check whether you need to install BioPerl. <br>
    <code>
    perl -MBio::SeqIO -e 0 <br>
    </code>
    If you get no feedback, BioPerl is available on your system. If you get an error message something like the following:<br>
    <code>
    Can't locate Bio/SeqIO.pm in @INC (@INC contains: /etc/perl /usr/local/lib/perl/5.14.2 ...    <br>
    </code>
    this indicates that you need to install BioPerl. <br><br>
    To install BioPerl, go to <a href="http://www.bioperl.org/wiki/Getting_BioPerl" title="Link to BioPerl website">the BioPerl Wiki</a> and follow the instructions given there. <br>
    </p>
   <li><b>Scripts</b></li>
   <p>Our scripts for Tag-Seq analysis are hosted on GitHub. You'll want to download scripts from this repository:
   <ul>
    <li><a href="https://github.com/Eli-Meyer/TagSeq_utilities/" title="Link to GitHub repository">Scripts specifically for Tag-Seq analysis.<a></li>
   </ul>
   <br>The best way to do this is to use the <i>git</i> tool. While this is undoubtedly the best and most widely-used platform<br>
   for sharing code, the git tool itself is not simple or intuitive. To learn about use of git, we recommend
   <a href="http://gitref.org/" title="Link to GitRef website">this resource</a>.<br>
   To download the above repositories onto your system, assuming <i>git</i> is available:<br>
   <code>git clone git://github.com/Eli-Meyer/TagSeq_utilities.git<br></code>
   Make sure you have execute permission to these scripts, and that they are in your path ($PATH).<br>
   </p>
   <li><b>bbduk.sh</b></li>
   <p>To check whether bbduk is installed on your system, run:<br>
   <code>which bbduk.sh</code><br>
   If you get no feedback, this indicates you need to install bbduk. Go to this
   <a href="https://sourceforge.net/projects/bbmap/" title="Link to bbmap website">link</a>
   and follow the instructions to unpack this collection of software, which requires Java version 6 or higher is installed on your system. <br>
   If you get feedback similar to this:<br>
   <code>/local/cluster/bbmap/bbduk.sh</code><br>
   This indicates bbduk is already installed and in your path. You can move on to the next step.<br>
   </p>
   <li><b>SHRiMP</b></li>
   <p>To check whether SHRiMP is installed on your system, run:<br>
   <code>which gmapper</code><br>
   If you get no feedback, this indicates you need to install SHRiMP. Go to this
   <a href="http://compbio.cs.toronto.edu/shrimp/" title="Link to SHRiMP website">link</a>
   and download and install SHRiMP according to the directions on that site.<br>
   If you get feedback similar to this:<br>
   <code>/local/cluster/SHRiMP/bin/gmapper</code><br>
   This indicates SHRiMP is already installed and in your path. You can move on to the next step.<br>
   </p>
   </ul>
   </div>
  </div>

  <div class="row"><br>
  <a href="#process" class="hide" id="process" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Trim and filter reads</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Trim and filter reads</b></span></a>
   <div class="list">
   <ul>
    <li><b>Quality Filter</b></li>
   <p>First, we exclude low quality reads likely to contain sequencing errors. The choice of thresholds is
   somewhat arbitrary and ultimately up to the user. For an example using reasonable default settings, run:<br>
   <code>QualFilterFastq.pl -i input.fastq -m 20 -x 10 -o output.fastq</code><br>
   to remove any reads from "input.fastq" having more than 10 bases with quality scores lower than 20, and write
   the output to a file named "output.fastq".<br>
   </p>
    <li><b>Homopolymer repeat filter Filter</b></li>
   <p>Some sequence datasets include lots of homopolymer repeats (long stretches of a single nucleotide, e.g. poly-A tails). 
    Generally this is a challenge for transcriptome assembly, and to a lesser extent for gene expression analysis (RNA-Seq) and 
    genome assembly, and not at all for SNP genotyping. To exclude all reads with homopolymer repeats more than 30 bases in length, 
    we can run:<br>
   <code>HRFilterFastq.pl -i input.fastq -n 30 -o output.fastq</code><br>
   </p>
    <li><b>Adaptor filter</b></li>
   <p>Next, we discard reads containing sequences derived from adaptors used during library preparation. These
    artificial DNA sequences introduced during library preparation may occupy unknown portions of the read 
    (including the entire read). Removing these is probably the most important task in read processing, and certainly the most
   computationally intensive. To make the whole pipeline easier to run in parallel, we've recently
   switched from the old version (AdaptorFilterFastq.pl) to a newer kmer-based tool
   called <a href="http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide"
   title="Documentation for BBDUK">BBDUK<a/>. </br>

   This has superior performance in all respects to the old process and most importantly, can be run
   directly on large sequence datasets without lots of tedious parallelizing.<br><br>

   To run this tool on a set of reads called input.fastq, to eliminate any reads sharing at least
   one 12-mer with sequences in adaptors.fasta, we can run:<br>
   <code>bbduk.sh in=hrf.fastq ref=adaptors.fasta k=12 stats=stats.txt out=clean.fastq overwrite=t</code><br>
   to remove any reads in "input.fastq" having valid alignments (at least one 12-bp kmer match)
   to sequences in "adaptors.fasta", and write the output (sequences passing the filter) to "output.fastq".<br>
    
    <li><b>Remove PCR Duplicates</b></li>
   <p>In this optional step, the user can choose to discard PCR duplicates based on the sequence of the insert and the sequence
    of an internal barcode incorporated during FS-cDNA synthesis. To remove duplicates from a library having barcodes at positions 1-4 
    and insert sequences beginning at position 9, the user would run:<br>
    <code>RemovePCRDups.pl -i input.fastq -o output.fastq -s 1 -e 4 -j 9</code>
   </p>
    
    <li><b>Trim tags</b></li>
   <p>Finally, we remove the non-template bases introduced at the 5' end of each tag during library preparation.
   The region removed begins at the 5' end by default (position 1), and extends to the end of the 5'-most occurence
    of GGG (up to a maximum position defined by the user). This range is determined from the position of the GGG motif in 
    RNA oligos used during library preparation. To trim a set of reads where the GGG motif is expected to occur in the first 8 bases:<br>
   <code>TagTrimmer.pl -i input.fastq -b 1 -e 8 -o output.fastq</code><br>
      </p>
   </ul>
   </div>
  </div>

    <div class="row"><br>
  <a href="#align" class="hide" id="align" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Align reads against reference</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Align reads against reference</b></span></a>
   <div class="list">
   <ul>
   <li><b>Align reads to reference</b></li>
   <p>The ideal reference for mapping Tag-Seq reads would be a complete and non-redundant set of transcripts from the same
    species as the reads. <i>de novo</i> transcriptome assemblies and transcripts identified from genome assemblies are often
    used for this purpose.</p>
    <p>We use the SHRiMP software package for mapping, which is both faster and more sensitive in our testing than
   any other mapping software we have tested. However, in principle other mapping programs can be substituted
   at this step, as long as the output is in SAM format and includes positive alignment scores in the "AS:" field.
   To use SHRiMP, run the <i>gmapper</i> tool, as:<br>
   <code>gmapper --qv-offset 33 -Q --strata -o 3 -N 1 reads.fastq reference.fasta >gmapper.sam</code><br>
   In this example, a set of trimmed and filtered reads ("reads.fastq") is mapped against a reference sequence ("reference.fasta")
   using a single thread (processor). The raw alignments are written to a file called "gmapper.sam". <br>
   <li><b>Filter alignments</b></li>
   The raw alignments may include weak and ambiguous alignments (i.e. cases where a single read matches equally well
   to two or more regions of the genome). To remove these:<br>
   <code>SAMFilter.pl -i gmapper.sam -m 40 -o filtered.sam -c 1</code><br>
   In this example, we chose to exclude any alignments having fewer than 40 matching bp. Ambiguous matches are also eliminated 
   by default. The alignments passing this filter are written to an output file called "filtered.sam". The counts data (the number of reads
     matching each gene) is reported in "counts.tab"<br>
   </ul>
<a href="#combine" class="hide" id="combine" style="text-decoration: none; color: black; margin-left: 2%; font-size:14pt">
   <span title="Click to expand">
   <b>&#9655 Combine expression data</b></span></a>
  <a href="#top" class="show" id="top" style="text-decoration: none; color: blue; margin-left: 2%; font-size:14pt">
   <span title="Click to collapse">
   <b>&#9665 Combine expression data</b></span></a>
   <div class="list">
   <ul>
   <li><b>Combine counts files from all samples</b></li>
   <p>Finally, we combine the counts data from all samples into a single output file, with samples as columns and genes as rows.</p>
   To combine data from three specific samples, you could run:<br>
    <code>CombineExpression.pl sample1/counts.tab sample2/counts.tab sample3/counts.tab > combined.tab</code>
   To combine data from all samples contained in subdirectories of the working directory:<br>
   <code>CombineExpression.pl */counts.tab > combined.tab</code>
   <p>The combined file produced by this step is ready for statistical analysis in DESeq2, edgeR, or other R packages for 
    analysis of differential gene expression.<br></p>
   </ul>
   </div>
  </div>

    
 <hr align="left" style="margin-left: 2%;">
  <p style="margin-left: 2%; font-size: 85%;">Last updated 14 July 2018, E. Meyer.</p>

 </body>
</html>
